collect/cache/cuckoo.go

```diff
@@ -24,32 +26,78 @@ const (
 // This is why the future filter is nil until the current filter reaches .5.
 // You must call Maintain() periodically, most likely from a goroutine. The call is cheap,
 // and the timing isn't very critical. The effect of going above "capacity" is an increased
-// false positive rate, but the filter continues to function.
+// false positive rate and slightly reduced performance, but the filter continues to function.
 type CuckooTraceChecker struct {
 	current  *cuckoo.Filter
 	future   *cuckoo.Filter
 	mut      sync.RWMutex
 	capacity uint
 	met      metrics.Metrics
+	addch    chan string
 }
 
+const (
+	// This is how many items can be in the Add Queue before we start blocking on Add.
+	AddQueueDepth = 10000
+	// This is the max number of items we'll pull from the queue during one lock cycle.
+	AddQueueBatchSize = 5
+	// This is how long we'll sleep between possible lock cycles.
+	AddQueueSleepTime = 100 * time.Microsecond
+)
+
 func NewCuckooTraceChecker(capacity uint, m metrics.Metrics) *CuckooTraceChecker {
-	return &CuckooTraceChecker{
+	c := &CuckooTraceChecker{
 		capacity: capacity,
 		current:  cuckoo.NewFilter(capacity),
 		future:   nil,
 		met:      m,
+		addch:    make(chan string, AddQueueDepth),
 	}
+
+	// To avoid blocking on Add, we have a goroutine that pulls from a channel and adds to the filter.
+	traces := make([]string, AddQueueBatchSize)
+	go func() {
+		for {
+			ready := false
+			i := 0
```

Comment:
I think `i` needs instantiated outside the `for` loop and then reset inside `if ready && i > 0 {`